### PR TITLE
Multi-level masking

### PR DESCRIPTION
**Pull request by 51N84D**
Current approach just adds bit-conditioning to the latent vector, and determines which (simulated) ground truth mask to compute losses with.

### CODE REVIEW

**Comment by vict0rsch** [comment]
Is that up to date? What's keeping us from merging?

**Comment by 51N84D** [comment]
Alright it's finally ready to review. Try running it before approving, just to make sure nothing is broken...

File: omnigan/trainer.py [hunk 1]
```diff
@@ -221,13 +220,35 @@ def setup(self):
...
+
         self.input_shape = self.compute_input_shape()
         if "s" in self.opts.tasks:
             self.G.decoders["s"].set_target_size(self.input_shape[-2:])
 
         print("OK.")
         self.painter_z_h = self.input_shape[-2] // (2 ** self.opts.gen.p.spade_n_up)
         self.painter_z_w = self.input_shape[-1] // (2 ** self.opts.gen.p.spade_n_up)
+        if "m2" in self.opts.tasks:
+            self.label_1 = torch.zeros(
```
Comment (vict0rsch) [inline omnigan/trainer.py:235]: maybe `label_0` and `label_1` or `label_low` `label_high` would make more sense

**Review by zhangtyu — APPROVED** [review]

File: omnigan/trainer.py [hunk 2]
```diff
@@ -441,8 +462,24 @@ def log_comet_images(self, mode, domain):
...
+                                dim=1,
+                            )
+                        )
 
-                    prediction = self.G.decoders[update_task](self.z)
+                    elif update_task == "m":
+                        if "m2" in self.opts.tasks:
+                            prediction = self.G.decoders[update_task](
+                                torch.cat(
+                                    (self.z, self.label_1[0, :, :, :].unsqueeze(0)),
```
Comment (vict0rsch) [inline omnigan/trainer.py:477]: idem

File: omnigan/trainer.py [hunk 3]
```diff
@@ -441,8 +462,24 @@ def log_comet_images(self, mode, domain):
 
                     if update_task not in save_images:
                         save_images[update_task] = []
+                    if update_task == "m2":
+                        prediction = self.G.decoders["m"](
+                            torch.cat(
+                                (self.z, self.label_2[0, :, :, :].unsqueeze(0)),
```
Comment (vict0rsch) [inline omnigan/trainer.py:468]: self.label_0[0, :, :, :].unsqueeze(0) should be the same as self.label_0[:1, ...]

File: omnigan/trainer.py [hunk 4]
```diff
@@ -457,24 +494,22 @@ def log_comet_images(self, mode, domain):
...
                         task_saves.append(target)
-
-                    elif update_task == "m":
+                    if update_task == "m" or update_task == "m2":
                         prediction = prediction.repeat(1, 3, 1, 1)
                         task_saves.append(x * (1.0 - prediction))
                         task_saves.append(x * (1.0 - target.repeat(1, 3, 1, 1)))
 
-                    elif update_task == "d":
+                    if update_task == "d":
```
Comment (vict0rsch) [inline omnigan/trainer.py:502]: why not `elif` ?

File: omnigan/trainer.py [hunk 5]
```diff
@@ -772,64 +808,141 @@ def get_masker_loss(self, multi_domain_batch):  # TODO update docstrings
...
+                                * self.opts.train.lambdas.G.m2.aux
+                            )
+
+                            self.logger.losses.generator.task_loss["m"]["aux"][
+                                batch_domain
+                            ] = update_loss.item()
+
+                            step_loss += update_loss
+
+                        elif update_task == "m2":
```
Comment (vict0rsch) [inline omnigan/trainer.py:852]: If I read this right, the only things that change between the `if` and `elif` are `self.label_1[:, 0, 0, 0].squeeze()` vs `self.label_2[:, 0, 0, 0].squeeze()` and `self.logger.losses.generator.task_loss`. I bet you can refactor this whole block in a much shorter way by having variables and a common code dependent on those. Cleaner, shorter, less error-prone (don't need to change 2 pieces of code if you change the logic) more versatile (what about more flood levels?)

File: omnigan/trainer.py [hunk 6]
```diff
@@ -1073,20 +1189,148 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
...
-                        prob = torch.cat([fake_mask, fake_complementary_mask], dim=1)
-                        prob = prob.detach()
 
-                        loss_main = self.losses["D"]["advent"](
-                            prob.to(self.device),
-                            self.domain_labels[batch_domain],
-                            self.D["m"]["Advent"],
-                        )
+                        if "m2" in self.opts.tasks:
+                            # --------ADVENT LOSS---------------
```
Comment (vict0rsch) [inline omnigan/trainer.py:1194]: Add `CoBlock` to your vscode extension and use cmd+shift+k to make nice comment blocks instead of those atrocious imbalanced hybrid headers

Comment (vict0rsch) [reply omnigan/trainer.py:1194]: :p

**Review by vict0rsch — CHANGES_REQUESTED** [review]
Requesting changes because I'd like to go over some of the code liv with you @51N84D

File: omnigan/trainer.py (hunk 1 above)
Comment (melisandeteng) [reply omnigan/trainer.py:235]: can we actuallly put low high ?

File: omnigan/discriminator.py [hunk 7]
```diff
@@ -281,6 +391,21 @@ def __init__(self, opts):
                     )
                 else:
                     raise Exception("This Discriminator is currently not supported!")
+        if "m2" in opts.tasks:
+            # Create a flood-level discriminator / classifier
+            self["m2"] = nn.ModuleDict(
```
[the remark this answers is not in the document]
Comment (51N84D) [reply omnigan/discriminator.py:396]: This seems to be what we're doing for tasks 'm' and 's' as well, even though they each only have one key ("Advent"). This leaves the door open to adding more discriminators if we want to right?

File: omnigan/trainer.py (hunk 1 above)
Comment (51N84D) [reply omnigan/trainer.py:235]: I kept it as numbers so that we can add more flood levels as needed. But I agree the lowest level should be label 0. I was anticipating bringing in the ground as label 0 (coming soon in an update)

Comment (vict0rsch) [reply omnigan/trainer.py:235]: oh ok then makes sense, I'm fine with it

**Comment by vict0rsch** [comment]
How about this PR, is it ready @51N84D ?
